cmd/src/actions_exec_backend_runner.go

```diff
@@ -0,0 +1,307 @@
+package main
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/shurcooL/go/ctxhttp"
+	"github.com/sourcegraph/go-diff/diff"
+)
+
+type ActionRepoStatus struct {
+	Cached bool
+
+	LogFile    string
+	EnqueuedAt time.Time
+	StartedAt  time.Time
+	FinishedAt time.Time
+
+	Patch CampaignPlanPatch
+	Err   error
+}
+
+func (x *actionExecutor) do(ctx context.Context, repo ActionRepo) (err error) {
+	// Check if cached.
+	cacheKey := actionExecutionCacheKey{Repo: repo, Runs: x.action.Run}
+	if result, ok, err := x.opt.cache.get(ctx, cacheKey); err != nil {
+		return errors.Wrapf(err, "checking cache for %s", repo.Name)
+	} else if ok {
+		x.updateRepoStatus(repo, ActionRepoStatus{
+			Cached: true,
+			Patch:  result,
+		})
+		return nil
+	}
+
+	prefix := "action-" + strings.Replace(strings.Replace(repo.Name, "/", "-", -1), "github.com-", "", -1)
+
+	// TODO(sqs): better cleanup of old log files
+	logFile, err := ioutil.TempFile("", prefix+"-log")
+	if err != nil {
+		return err
+	}
+	if !x.opt.keepLogs {
+		defer func() {
+			if err == nil {
+				os.Remove(logFile.Name())
+			}
+		}()
+	}
+	logWriter := io.Writer(logFile)
+	if *verbose {
+		logWriter = io.MultiWriter(logWriter, os.Stderr)
+	}
+
+	x.updateRepoStatus(repo, ActionRepoStatus{
+		LogFile:   logFile.Name(),
+		StartedAt: time.Now(),
+	})
+
+	// TODO!(sqs)
+	//
+	// ctx, cancel := context.WithTimeout(ctx, 90*time.Second)
+	// defer cancel()
+
+	patch, err := runAction(ctx, prefix, repo.ID, repo.Name, repo.Rev, x.action.Run, logWriter)
+	status := ActionRepoStatus{
+		FinishedAt: time.Now(),
+	}
+	if len(patch) > 0 {
+		status.Patch = CampaignPlanPatch{
+			Repository:   repo.ID,
+			BaseRevision: repo.Rev,
+			Patch:        string(patch),
+		}
+	}
+	if err != nil {
+		status.Err = err
+		fmt.Fprintf(logWriter, "# ERROR: %s\n", err)
+	}
+	x.updateRepoStatus(repo, status)
+
+	// Add to cache if successful.
+	if err == nil {
+		if err := x.opt.cache.set(ctx, cacheKey, status.Patch); err != nil {
```

Comment:
Note to myself: does this also write to the cache when `status.Patch == CampaignPlanPatch{}`? Or do we need the `len(patch) > 0` check here too?